In [1]:
import pandas as pd
import numpy as np

### 데이터셋
- 일단 빠른 작업 시작을 위해, `mh_PHQ_S`변수가 수집된 2014, 16, 18, 20 년도 DB를 받아 취합한 csv를 올려서 사용했습니다. 

In [2]:
df_origin = pd.read_csv("./HN_14_16_18_20.csv")

/tmp/ipykernel_202247/1157134730.py:1: DtypeWarning: Columns (2,42,43,143,148,192,199,203,210,217,260,268,289,347,358,399,556,573,595,600,601,747,758,759,786,801,803,812,823,847,870,876,881,891,913,915,920,921,922,923,924,925,926,928,930,987,991,994,995,1071) have mixed types. Specify dtype option on import or set low_memory=False.
  df_origin = pd.read_csv("./HN_14_16_18_20.csv")


In [3]:
# change column name 
df_origin.rename(columns={'Unnamed:0': 'index'}, inplace=True)

In [4]:
df_origin.info() # shape: (31051, 1107)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31051 entries, 0 to 31050
Columns: 1107 entries, Unnamed: 0 to DA_YOGURT
dtypes: float64(1047), int64(1), object(59)
memory usage: 262.2+ MB


In [5]:
# Check overall missing values -> too many columns to see clearly hmm
df_origin.isnull().sum()

Unnamed: 0        0
mod_d             0
id            23501
ID_fam            0
year              0
              ...  
DA_MILK       25242
DQ_SMILK      25242
DA_SMILK      25242
DQ_YOGURT     25242
DA_YOGURT     25242
Length: 1107, dtype: int64

In [6]:
features = ['mh_PHQ_S','DF2_pr',
 'BP_PHQ_1',
 'BP_PHQ_2',
 'BP_PHQ_3',
 'BP_PHQ_4',
 'BP_PHQ_5',
 'BP_PHQ_6',
 'BP_PHQ_7',
 'BP_PHQ_8',
 'BP_PHQ_9',
 'DK8_dg',
 'BO3_05',
 'HE_Uket',
 'HE_fh',
 'BA2_12',
 'LQ4_03',
 'LQ4_15',
 'LQ4_21',
 'HE_Usg',
 'HE_BMI',
 'BE3_82',
 'DQ_MILK',
 'HE_STRfh1',
 'BM2_1',
 'BD2',
 'T_Q_DZ6',
 'E_CLT',
 'T_NQ_OCP',
 'BM7',
 'BM13_3',
 'E_CLC',
 'LW_pr_1',
 'D_2_wk',
 'DM4_pr',
 'DC3_pr',
 'BS12_1',
 'BE3_72',
 'T_Q_DZ5',
 'BP16_11',
 'L_OUT_FQ',
 'DJ8_pr',
 'LQ2_mn',
 'DI1_dg',
 'BP6_10',
 'HE_Uro',
 'DJ4_pr',
 'BM2_3',
 'DH3_dg',
 'HE_BUN',
 'HE_cough2',
 'E_EM',
 'DM8_pt',
 'HE_LDL_drct',
 'EC1_1',
 'LQ_1EQL',
 'LQ4_08',
 'DQ_SODA',
 'incm',
 'BA2_2_3',
 'LQ4_06',
 'DC5_pr',
 'LQ4_04',
 'E_PEO_1',
 'BS9_2',
 'BH9_11',
 'house',
 'HE_HTG',
 'T_Q_SNST2',
 'DK4_pr',
 'LW_mt',
 'BA2_2_2',
 'T_IM_EX_L_1',
 'MO4_18',
 'BS6_2',
 'BE3_85',
 'HE_sbp',
 'L_DN_TO',
 'BM1_7',
 'HE_STRfh2',
 'DE2_pr',
 'BO1_1',
 'HE_chol',
 'T_Q_CR',
 'HE_HLfh2',
 'LQ4_11',
 'HE_df',
 'ID_fam',
 'DH4_dg',
 'LQ4_00',
 'HE_HBfh3',
 'HE_sput1',
 'LQ1_mn',
 'DC1_pt',
 'EC_stt_1',
 'E_DR',
 'HE_Uglu',
 'HE_prg',
 'HE_fev1p',
 'DK8_pt',
 'DH7_pt',
 'EC_occp',
 'DC11_pt',
 'DJ2_pr',
 'HE_DM_HbA1c',
 'BM1_5',
 'ID',
 'BM13',
 'LQ4_01',
 'E_DG',
 'HE_THfh2',
 'MH1_1',
 'BS12_2',
 'BP16_13',
 'BA2_2_4',
 'DL1_pr',
 'E_CL_pr',
 'ainc_1',
 'O_ortho',
 'DI2_pr',
 'ainc',
 'DC1_pr',
 'DM1_pr',
 'L_BR_WHO',
 'HE_HP',
 'BE8_1',
 'HE_pef',
 'graduat',
 'LQ4_23',
 'T_VHIT_EX_L_1',
 'allownc',
 'OR1_2',
 'E_NWT',
 'BE3_91',
 'BM1_1',
 'BP1',
 'LQ4_13',
 'genertn',
 'DH7_dg',
 'DM8_dg',
 'LW_oc',
 'LQ4_29',
 'LQ4_09',
 'E_CLD',
 'BO3_10',
 'LK_LB_US',
 'E_RT',
 'DC7_pt',
 'BA2_14',
 'BA2_13',
 'year',
 'BH1',
 'BM2_2',
 'HE_obe',
 'ho_incm',
 'DC12_pr',
 'BO3_11',
 'HE_Unitr',
 'DC5_dg',
 'DM4_pt',
 'E_Cr_1',
 'DM3_pt',
 'HE_HDL_st2',
 'LW_ms_a',
 'LQ4_16',
 'BE3_32',
 'T_Q_VN_1',
 'BM13_1',
 'BS5_1',
 'DC6_pt',
 'DI3_pr',
 'HE_hepaC',
 'HE_dog',
 'HE_DMfh3',
 'BP5',
 'DJ4_3',
 'HE_HbA1c',
 'HE_hepaB',
 'live_t',
 'DI2_pt',
 'DH4_pt',
 'LQ_4EQL',
 'DH3_pr',
 'DK4_dg',
 'BE3_93',
 'BM14',
 'LQ_4HT',
 'HE_HCT',
 'DJ6_pr',
 'D_2_1',
 'sex',
 'DI2_2',
 'LW_mt_a2',
 'EC_pedu_2',
 'BO3_14',
 'HE_Uph',
 'L_BR_TO',
 'DC3_pt',
 'DC12_pt',
 'occp',
 'HE_cough1',
 'HE_ht',
 'DQ_SMILK',
 'BD1',
 'LQ4_22',
 'HE_wc',
 'E_NMRC_r',
 'BE5_1',
 'HE_HLfh3',
 'T_VHIT_GA_R',
 'HE_cat',
 'BO3_03',
 'BO3_12',
 'T_Q_HR',
 'BP7',
 'HE_wt',
 'LQ4_10',
 'BS3_3',
 'LQ_5EQL',
 'HE_fev6',
 'LK_LB_EF',
 'LW_ms',
 'O_chew_d',
 'LQ4_26',
 'LS_FRUIT',
 'E_Pr_1',
 'cfam',
 'HE_DMfh2',
 'BM1_4',
 'HE_STRfh3',
 'BS12_37',
 'BA2_2_6',
 'DC11_dg',
 'D_1_1',
 'HE_DMdr',
 'HE_ast',
 'T_Q_VN',
 'DN1_dg',
 'DH7_pr',
 'MO4_17_2',
 'EC_wht_5',
 'E_CT',
 'E_NMRC_l',
 'DQ_CAFFEINE',
 'T_Q_ALLE',
 'HE_HPfh2',
 'DK4_pt',
 'DK9_dg',
 'BD7_62',
 'LQ4_07',
 'L_LN_WHO',
 'BE3_83',
 'HE_PFThs',
 'MO4_8',
 'T_VHIT_EX_R',
 'HE_ragweed',
 'E_MH',
 'E_DR1',
 'HE_RBC',
 'E_GR',
 'T_Q_DZ3',
 'DI4_pt',
 'HE_HPfh3',
 'BS12_47',
 'DJ2_pt',
 'DK9_pr',
 'LQ_7HT',
 'DC11_pr',
 'LQ4_14',
 'DJ6_dg',
 'T_Q_DZ4',
 'DM2_pt',
 'HE_fvcp',
 'E_Pl_1',
 'EC_wht_23',
 'E_Dl_2_1',
 'T_VHIT_COR_R',
 'LW_br',
 'LQ2_ab',
 'HE_fev1',
 'BP16_23',
 'BE3_86',
 'HE_sput2',
 'DN1_pr',
 'HE_HBfh1',
 'T_Q_SNST1',
 'E_CLC_T',
 'DM8_pr',
 'DC6_pr',
 'DM3_pr',
 'DJ4_dg',
 'HE_glu',
 'HE_HPfh1',
 'DI1_2',
 'HE_THfh3',
 'DC1_dg',
 'DI3_pt',
 'T_Q_CR3',
 'MO4_00',
 'T_VHIT_GA_L',
 'T_Q_HR_1',
 'BM2_5',
 'HE_IHDfh3',
 'LW_mt_a1',
 'BE3_75',
 'DQ_YOGURT',
 'E_DME',
 'BA2_2_1',
 'E_CLU',
 'educ',
 'L_LN_TO',
 'LQ4_12',
 'BA2_2_5',
 'DJ8_pt',
 'DE2_pt',
 'L_BR_FQ',
 'HE_Uacid',
 'marri_1',
 'DE1_pt',
 'EC_stt_2',
 'HE_Bplt',
 'HE_birch',
 'E_Cl_1',
 'DC4_dg',
 'BD2_1',
 'DC12_dg',
 'HE_fev1fvc',
 'DC4_pt',
 'BE3_31',
 'BS6_3',
 'T_IM_EX_R_1',
 'DC4_pr',
 'DM4_dg',
 'DC6_dg',
 'MO4_15',
 'LS_VEG1',
 'DH3_pt',
 'HE_HCHOL',
 'DC2_pr',
 'T_Q_CR4',
 'DQ_TEA',
 'ainc_unit1',
 'T_VHIT_EX_R_1',
 'HE_Ubld',
 'T_Q_VC',
 'HE_dbp',
 'BS6_4',
 'T_IM_EX_R',
 'BO3_09',
 'BM1_6',
 'BO1',
 'BS1_1',
 'LQ_6HT',
 'DE1_pr',
 'DJ8_dg',
 'BD7_65',
 'MO4_17_1',
 'T_Q_DZ',
 'HE_Upro',
 'T_ex',
 'LQ_2HT',
 'HE_WBC',
 'BO3_02',
 'DH6_pr',
 'E_DG1',
 'HE_TG',
 'HE_mens',
 'T_IM_EX_L',
 'BH2_61',
 'BD1_11',
 'DI4_pr',
 'BD7_5',
 'E_AMD',
 'BO2_1',
 'DQ_COFFEE',
 'DC2_pt',
 'DL1_pt',
 'DN1_pt',
 'T_Q_VN_2',
 'MO4_17',
 'HE_oak',
 'DJ4_pt',
 'BE3_71',
 'BP6_2',
 'L_DN_WHO',
 'DK8_pr',
 'DH6_dg',
 'MO4_9',
 'T_Q_VC1',
 'DI3_dg',
 'DH2_pr',
 'BM13_4',
 'HE_DMfh1',
 'DI4_dg',
 'HE_fvc',
 'DJ2_dg',
 'DI1_pt',
 'M_2_yr',
 'BE3_77',
 'HE_fef25_75',
 'BO3_01',
 'DK9_pt',
 'DM1_dg',
 'BM1_3',
 'BE3_92',
 'HE_HBfh2',
 'BM2_4',
 'HE_HLfh1',
 'HE_PFTtr',
 'BM1_8',
 'DI2_dg',
 'marri_2',
 'DC3_dg',
 'LQ4_25',
 'age',
 'HE_IHDfh2',
 'HE_PFTdr',
 'MO4_12',
 'T_Q_DZ7',
 'T_Q_SINU',
 'BP16_2',
 'EC_pedu_1',
 'LW_mp_a',
 'BP16_1',
 'BP16_21',
 'BS5',
 'HE_Ualb',
 'BP16_12',
 'BS3_2',
 'HE_Ucot',
 'E_DM',
 'BO3_04',
 'LQ_2EQL',
 'DH4_pr',
 'T_Q_CR2',
 'DQ_STEA',
 'T_VHIT_COR_L',
 'BD7_61',
 'LQ_3EQL',
 'ho_incm5',
 'BM13_2',
 'E_PEO',
 'BS12_47_1',
 'BM8',
 'BS13',
 'HE_COPD',
 'HE_nc',
 'DC7_pr',
 'LW_pr',
 'L_DN_FQ',
 'DH6_pt',
 'BH2_62',
 'T_Q_SNST3',
 'LQ1_sb',
 'T_Q_SINU_1',
 'DM2_dg',
 'BE3_87',
 'DC7_dg',
 'BE3_73',
 'HE_Ubil',
 'incm5',
 'HE_crea',
 'pa_aerobic',
 'T_Q_DZ2',
 'DI1_pr',
 'MO4_7',
 'LK_LB_CO',
 'BO1_3',
 'HE_HPdr',
 'HE_PFTag',
 'HE_insulin',
 'HE_IHDfh1',
 'BM1_2',
 'DI3_2',
 'L_LN_FQ',
 'BS3_1',
 'BD7_64',
 'LQ_8HT',
 'HE_UNa',
 'T_Q_CR1',
 'T_VHIT_EX_L',
 'DJ6_pt',
 'DH2_pt',
 'BS5_4',
 'LQ_5HT',
 'BM13_5',
 'E_EXC_1',
 'HE_Ukal',
 'LQ_1HT',
 'DM3_dg',
 'BD7_63',
 'DM2_pr',
 'LQ4_05',
 'DE1_dg',
 'MO4_11',
 'BM14_1',
 'T_Q_SNST4',
 'HE_Ucrea',
 'BE3_81',
 'HE_anem',
 'MH1_yr',
 'HE_Jhop',
 'DL1_dg',
 'DC5_pt',
 'O_pain',
 'LS_VEG2',
 'HE_THfh1',
 'BO3_07',
 'E_ex',
 'DE2_dg',
 'DQ_SCOFFEE',
 'LQ4_28',
 'DH2_dg',
 'LQ_3HT',
 'E_RVO',
 'HE_rPLS',
 'BO1_2',
 'DM1_pt',
 'LQ4_02',
 'DC2_dg',
 'HE_dprg',
 'HE_alt',
 'E_Q_EX',
 'E_Dr_2_1',
 'BD7_66',
 'T_Q_HR_2',
 'BE3_76',
 'MO4_4',
 'EC_wht_0',
 'BP6_31',
 'fam_rela',
 'LQ4_27',
 'BS8_2',
 'BS2_2',
 'HE_HB',
 'edu']

In [7]:
# drop un-existed features in this dataset
exclude_lst = ['HE_df', 'T_VHIT_EX_L_1', 'E_RT', 'HE_dog', 'LQ_4HT', 'T_VHIT_GA_R', 'HE_cat', 'T_VHIT_EX_R', 'HE_ragweed', 'LQ_7HT', 'T_VHIT_COR_R', 'E_CLC_T', 'T_VHIT_GA_L', 'HE_birch', 'T_VHIT_EX_R_1', 'LQ_6HT', 'LQ_2HT', 'HE_oak', 'T_VHIT_COR_L', 'LQ_8HT', 'T_VHIT_EX_L', 'LQ_5HT', 'LQ_1HT', 'HE_Jhop', 'LQ_3HT']

features = [x for x in features if x not in exclude_lst]
df = df_origin[features]

In [8]:
df.info() # shape: (31051, 511) # Including 'mh_PHQ_S','DF2_pr','BP_PHQ_1 ~ BP_PHQ_9'

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31051 entries, 0 to 31050
Columns: 511 entries, mh_PHQ_S to edu
dtypes: float64(504), object(7)
memory usage: 121.1+ MB


In [9]:
# Check missing values in y candidates 
df[['mh_PHQ_S','DF2_pr', 'BP_PHQ_1', 'BP_PHQ_2', 'BP_PHQ_3', 
    'BP_PHQ_4', 'BP_PHQ_5', 'BP_PHQ_6', 'BP_PHQ_7', 'BP_PHQ_8', 'BP_PHQ_9']].isnull().sum()

mh_PHQ_S    9074
DF2_pr      1629
BP_PHQ_1    1629
BP_PHQ_2    1629
BP_PHQ_3    1629
BP_PHQ_4    1629
BP_PHQ_5    1629
BP_PHQ_6    1629
BP_PHQ_7    1629
BP_PHQ_8    1629
BP_PHQ_9    1629
dtype: int64

### 데이터에서 청소년, 소아의 경우를 제외하기
- 만 19세 이상인 성인만을 대상으로 데이터를 다루기로 정의했기 때문에, `age`변수를 이용해서 청소년과 소아를 제외한다.

In [10]:
# Filter only adult data using `age` variable
df = df[df['age'] >= 19]

### 결측값 처리 - (1)
- row 기준, 모든 column의 응답에서 30% 이상을 결측치로 응답한 사람은 drop한다.
    - 10% 이상으로 잡으면 1627개만 남음
- column 기준 결측치를 drop할 필요가 지금 단계에서 있을까?

In [11]:
# threshold for null proportion
threshold = 0.3
null_percent = df.isnull().mean(axis = 1) # 각 행에 대한 결측값 비율 계산
df = df[null_percent < threshold]
df


mh_PHQ_S  DF2_pr  BP_PHQ_1  BP_PHQ_2  BP_PHQ_3  BP_PHQ_4  BP_PHQ_5  \
0           1.0     8.0       0.0       0.0       1.0       0.0       0.0   
3           0.0     8.0       0.0       0.0       0.0       0.0       0.0   
5           3.0     8.0       0.0       1.0       0.0       0.0       0.0   
8           9.0     8.0       1.0       1.0       1.0       1.0       0.0   
9           1.0     8.0       0.0       0.0       0.0       1.0       0.0   
...         ...     ...       ...       ...       ...       ...       ...   
31046       0.0     8.0       0.0       0.0       0.0       0.0       0.0   
31047       3.0     8.0       1.0       1.0       0.0       1.0       0.0   
31048       3.0     8.0       0.0       0.0       2.0       1.0       0.0   
31049       0.0     8.0       0.0       0.0       0.0       0.0       0.0   
31050       0.0     0.0       0.0       0.0       0.0       0.0       0.0   

       BP_PHQ_6  BP_PHQ_7  BP_PHQ_8  ...  BE3_76  MO4_4  EC_wht_0  BP6_31  \
0           0.0       0.0       0.0  ...     4.0    1.0       8.0     8.0   
3           0.0       0.0       0.0  ...     8.0    8.0       2.0     8.0   
5           1.0       0.0       0.0  ...     8.0    8.0       8.0     8.0   
8           3.0       0.0       0.0  ...     8.0    8.0       8.0     8.0   
9           0.0       0.0       0.0  ...     8.0    8.0       8.0     8.0   
...         ...       ...       ...  ...     ...    ...       ...     ...   
31046       0.0       0.0       0.0  ...     8.0    8.0       8.0     2.0   
31047       0.0       0.0       0.0  ...     2.0    1.0       2.0     2.0   
31048       0.0       0.0       0.0  ...     8.0    8.0       1.0     2.0   
31049       0.0       0.0       0.0  ...     8.0    1.0       8.0     2.0   
31050       0.0       0.0       0.0  ...     8.0    1.0       2.0     2.0   

       fam_rela  LQ4_27  BS8_2  BS2_2  HE_HB  edu  
0           3.0     NaN    8.0    NaN   13.9  4.0  
3           1.0     NaN    2.0    NaN   14.9  1.0  
5           2.0     NaN    8.0    NaN   15.3  4.0  
8           2.0     NaN    8.0    NaN   14.2  3.0  
9           1.0     NaN    1.0    NaN   15.0  4.0  
...         ...     ...    ...    ...    ...  ...  
31046       1.0     8.0    2.0    NaN   13.4  3.0  
31047       1.0     8.0    2.0    NaN   13.6  4.0  
31048       3.0     8.0    2.0    NaN   13.2  4.0  
31049       1.0     8.0    8.0    NaN   14.9  4.0  
31050       2.0     0.0    2.0    NaN   13.6  3.0  

[21114 rows x 511 columns]

### 우울증임을 정의하기 - y (dependent variable)
 `mh_PHQ_S` 변수란?  
	- 한글판 PHQ-9 (Patient Health Questionnaire-9) 검사를 평가 도구로 쓰고, 각 문항의 접수를 합산한 변수   
	- 문항은 수검자 본인이 직접 읽고 기입한 자기 기입식 설문 조사  
	- 국가건강검진에서 ``PHQ-9 검사 결과 9번 문항에서 1점 이상``으로 응답한 경우 및 ``총점이 10점 이상인 경우``, 일상생활에서 우울증을 극복할 수 있는 방법을 담은 <우울 증상과 극복 방법>을 제공한다. 또한 가까운 병/의원 또는 정신건강 복지센터에 의뢰해 신속하게 진료 및 치료를 받을 수 있도록 안내하고, 정신건강 위기 상담 전화를 통해 도움받을 수 있는 방법도 함께 안내한다.

- BP_PHQ_1 ~BP_PHQ_9은 null값이 1,629개인 반면 , mh_PHQ_S 는 null값이 9,074나 된다.
다만, BP_PHQ_1 ~BP_PHQ_9null값이 아니어도, 각 항목에 8(비해당), 9(모름/무응답)을 선택한 경우는 유효하지 않다.
따라서, 이미 생성된 mh_PHQ_S를 활용하되, BP_PHQ_9에서 1점 이상을 선택한 사용자도 선별해서 우울증 위험군으로 분류한다.

In [12]:
# Define depression (dependent variable)
cond1 = df['mh_PHQ_S'] >= 10 # total score 
cond2 = df['BP_PHQ_9'].isin([1, 2, 3]) # person who chose 1, 2, 3 in 9th question

depressed = df[cond1 | cond2].drop_duplicates() # person with cond1 or cond2

In [13]:
depressed.info() 

<class 'pandas.core.frame.DataFrame'>
Index: 1747 entries, 5 to 31042
Columns: 511 entries, mh_PHQ_S to edu
dtypes: float64(504), object(7)
memory usage: 6.8+ MB


In [14]:
# Create a new variable 'depressed' 
df.loc[cond1 | cond2, 'depressed'] = 1 # 두 조건 중 하나 이상을 만족하는 행에 'depressed' 변수에 1 값 
df.loc[~(cond1 | cond2), 'depressed'] = 0 # 두 조건 중 하나라도 만족하지 않는 행에 0 값

In [15]:
# Check distribution of depressed to see if the data is imbalanced
df.groupby('depressed').count() # groupby를 이용해 타겟 y 라벨의 클래스 별 개수를 확인

mh_PHQ_S  DF2_pr  BP_PHQ_1  BP_PHQ_2  BP_PHQ_3  BP_PHQ_4  BP_PHQ_5  \
depressed                                                                       
0.0           18234   19367     19367     19367     19367     19367     19367   
1.0            1721    1747      1747      1747      1747      1747      1747   

           BP_PHQ_6  BP_PHQ_7  BP_PHQ_8  ...  BE3_76  MO4_4  EC_wht_0  BP6_31  \
depressed                                ...                                    
0.0           19367     19367     19367  ...   19367  19367     19367   19367   
1.0            1747      1747      1747  ...    1747   1747      1747    1747   

           fam_rela  LQ4_27  BS8_2  BS2_2  HE_HB    edu  
depressed                                                
0.0           19367   16636  19367   5767  18949  18359  
1.0            1747    1426   1747    443   1684   1746  

[2 rows x 511 columns]

아래 표를 보면, `depressed`와 `DF2_pr`의 분포와 연관성을 한 눈에 확인할 수 있다.
먼저, `depressed` == 1 의 조건을 다시 생각해보자. 이는 수검자 본인이 응답한 결과에 따르면, 중증 우울증이 의심되는 조건이다. 
1) `mh_PHQ_S`, 즉 PHQ 설문의 총점이 10점 이상인 경우 혹은 (or)
2) `BP_PHQ_9`, PHQ-9 문항의 응답이 1점 이상인 경우이다. (자살 시도 혹은 계획)

`DF2_pr`은 우울증 현재 유병 여부를 나타내는 변수이다.
- `DF2_pr` == 8:은 전부 `DF2_dg`(우울증 의사진단 여부)에서 0(의사진단 받지 않음)을 선택한 사람이다. 
    -  `DF2_pr` == 8: `DF2)_dg`(우울증 의사진단 여부)에서 0(없음) 또는 8(비해당 - 청소년, 소아)을 선택한 경우인데, 이미 위에서 `age`를 이용해 데이터셋에 만 19세 이상인 성인만 남겨둔 상황이기 때문임. 


In [16]:
# Check distribution of who those are depressed or not with two variables `depressed` and `DF2_pr`
df.groupby(['depressed', 'DF2_pr']).count()

mh_PHQ_S  BP_PHQ_1  BP_PHQ_2  BP_PHQ_3  BP_PHQ_4  BP_PHQ_5  \
depressed DF2_pr                                                               
0.0       0.0          335       338       338       338       338       338   
          1.0          315       329       329       329       329       329   
          8.0        17584     17777     17777     17777     17777     17777   
          9.0            0       923       923       923       923       923   
1.0       0.0           70        70        70        70        70        70   
          1.0          287       293       293       293       293       293   
          8.0         1364      1384      1384      1384      1384      1384   

                  BP_PHQ_6  BP_PHQ_7  BP_PHQ_8  BP_PHQ_9  ...  BE3_76  MO4_4  \
depressed DF2_pr                                          ...                  
0.0       0.0          338       338       338       338  ...     338    338   
          1.0          329       329       329       329  ...     329    329   
          8.0        17777     17777     17777     17777  ...   17777  17777   
          9.0          923       923       923       923  ...     923    923   
1.0       0.0           70        70        70        70  ...      70     70   
          1.0          293       293       293       293  ...     293    293   
          8.0         1384      1384      1384      1384  ...    1384   1384   

                  EC_wht_0  BP6_31  fam_rela  LQ4_27  BS8_2  BS2_2  HE_HB  \
depressed DF2_pr                                                            
0.0       0.0          338     338       338     272    338     99    332   
          1.0          329     329       329     277    329     95    309   
          8.0        17777   17777     17777   15181  17777   5338  17436   
          9.0          923     923       923     906    923    235    872   
1.0       0.0           70      70        70      54     70     18     64   
          1.0          293     293       293     247    293     63    283   
          8.0         1384    1384      1384    1125   1384    362   1337   

                    edu  
depressed DF2_pr         
0.0       0.0       336  
          1.0       324  
          8.0     17698  
          9.0         1  
1.0       0.0        70  
          1.0       292  
          8.0      1384  

[7 rows x 510 columns]

### 학습에 방해가 될 수 있는 그룹 제외하기
일단, 잠재적으로 우울한 사람의 경우를 생각해보자.  
		- 1) 일단 의사에게 진단을 받은 적이 없음. (`DF2_pr == 8`)  
			- 1-1) `mh_PHQ_S == NaN` : PHQ 설문도 응답을 하지 않았기 때문에 이 사람들에 대해서는 아예 정보가 없음. -> drop 해야 함.  
			- 1-2) `mh_PHQ_S != NaN`: 이 사람들은 괜찮음. 어떻게던 분류가 될 것  
		- 2) `DF2_pr & mh_PHQ_S == NaN`: 우울증에 관련한 정보가 아예 없는 사람  
	- 위 2가지 케이스 모두 데이터셋에서 제외해야 함.   

In [27]:
# people who might be depressed
# DF2_pr == 0 & mh_PHQ_S == NaN

# 일단, mh_PHQ_S == NaN인 경우를 분리하고 생각하자 (나중에 drop할때에는 원래 df에서 제외해버리면 됨)
null_PHQ_S = df[pd.isnull(df['mh_PHQ_S'])]

# 1) DF2_pr == 8 (never been diagnosed by doctor) & mh_PHQ_S == NaN
null_PHQ_S[null_PHQ_S['DF2_pr'] == 8].shape # 213개 있음


(213, 512)

In [24]:
# 2) DF2_pr & mh_PHQ_S == NaN`: 우울증에 관련한 정보가 아예 없는 사람
null_PHQ_S[['DF2_pr', 'BP_PHQ_9']].isnull().sum() # DF2_pr, BP_PHQ_9(혹시 모르니까)도 null인 케이스를 세보자

DF2_pr      0
BP_PHQ_9    0
dtype: int64

In [28]:
# Remove people who might be depressed 
# DF2_pr == 8 (never been diagnosed by doctor) & mh_PHQ_S == NaN) 인 경우를 제외하려면?
# DF2_pr != 8 or mh_PHQ != NaN (전부 다 ~ 취하기)
df = df[(df['DF2_pr'] != 8 |  ~df['mh_PHQ_S'].isna() )] 
df

mh_PHQ_S  DF2_pr  BP_PHQ_1  BP_PHQ_2  BP_PHQ_3  BP_PHQ_4  BP_PHQ_5  \
0           1.0     8.0       0.0       0.0       1.0       0.0       0.0   
3           0.0     8.0       0.0       0.0       0.0       0.0       0.0   
5           3.0     8.0       0.0       1.0       0.0       0.0       0.0   
8           9.0     8.0       1.0       1.0       1.0       1.0       0.0   
9           1.0     8.0       0.0       0.0       0.0       1.0       0.0   
...         ...     ...       ...       ...       ...       ...       ...   
31046       0.0     8.0       0.0       0.0       0.0       0.0       0.0   
31047       3.0     8.0       1.0       1.0       0.0       1.0       0.0   
31048       3.0     8.0       0.0       0.0       2.0       1.0       0.0   
31049       0.0     8.0       0.0       0.0       0.0       0.0       0.0   
31050       0.0     0.0       0.0       0.0       0.0       0.0       0.0   

       BP_PHQ_6  BP_PHQ_7  BP_PHQ_8  ...  MO4_4  EC_wht_0  BP6_31  fam_rela  \
0           0.0       0.0       0.0  ...    1.0       8.0     8.0       3.0   
3           0.0       0.0       0.0  ...    8.0       2.0     8.0       1.0   
5           1.0       0.0       0.0  ...    8.0       8.0     8.0       2.0   
8           3.0       0.0       0.0  ...    8.0       8.0     8.0       2.0   
9           0.0       0.0       0.0  ...    8.0       8.0     8.0       1.0   
...         ...       ...       ...  ...    ...       ...     ...       ...   
31046       0.0       0.0       0.0  ...    8.0       8.0     2.0       1.0   
31047       0.0       0.0       0.0  ...    1.0       2.0     2.0       1.0   
31048       0.0       0.0       0.0  ...    8.0       1.0     2.0       3.0   
31049       0.0       0.0       0.0  ...    1.0       8.0     2.0       1.0   
31050       0.0       0.0       0.0  ...    1.0       2.0     2.0       2.0   

       LQ4_27  BS8_2  BS2_2  HE_HB  edu  depressed  
0         NaN    8.0    NaN   13.9  4.0        0.0  
3         NaN    2.0    NaN   14.9  1.0        0.0  
5         NaN    8.0    NaN   15.3  4.0        1.0  
8         NaN    8.0    NaN   14.2  3.0        1.0  
9         NaN    1.0    NaN   15.0  4.0        0.0  
...       ...    ...    ...    ...  ...        ...  
31046     8.0    2.0    NaN   13.4  3.0        0.0  
31047     8.0    2.0    NaN   13.6  4.0        0.0  
31048     8.0    2.0    NaN   13.2  4.0        0.0  
31049     8.0    8.0    NaN   14.9  4.0        0.0  
31050     0.0    2.0    NaN   13.6  3.0        0.0  

[20492 rows x 512 columns]

In [38]:
# Check distribution of depressed class - 0 or 1
df['depressed'].value_counts(normalize=True) # 0: 0.92, 1: 0.07

depressed
0.0    0.929045
1.0    0.070955
Name: proportion, dtype: float64

### 결측치 처리 - (2)
- column 기준으로 모든 행에 대한 결측값이 10% 이상이면 제외하기

In [42]:
# Remove columns if percentage of NaN exceeds 0.1
threshold = 0.1  

null_percent = df.isnull().mean() # 각 열에 대한 결측값 비율 계산
selected_cols = null_percent[null_percent < threshold].index
df_filtered = df[selected_cols]
df_filtered.shape # (20492, 314)

(20492, 314)

### Todo
- [ ] `depressed == 0 & DF2_pr == 1`인 사람들의 치료 여부(`DF2_pt`) 파악 
- [ ] 결측치 처리 (KNN 등)
- [ ] 이상치 확인
- [ ] feature 다시 정리해서 선택
- [ ] 너무 직접적인 영향을 주는 변수들 -> 확인 필요함